File: 1926-nearest-exit-from-entrance-in-maze/1926-nearest-exit-from-entrance-in-maze.py
# LarryNY's Code

class Solution:
    def nearestExit(self, maze: List[List[str]], entrance: List[int]) -> int:
        R = len(maze)
        C = len(maze[0])

        sx, sy = entrance
        q = collections.deque()
        INF = 10 ** 20
        distance = [[INF] * C for _ in range(R)]

        def enqueue(x, y, d):
            distance[x][y] = d
            q.append((x, y, d))

        enqueue(sx, sy, 0)

        directions = [(1, 0), (0, 1), (-1, 0), (0, -1)]
        while len(q) > 0:
            x, y, d = q.popleft()

            if x == 0 or x == R - 1 or y == 0 or y == C - 1:
                if not (x == sx and y == sy):
                    return d

            for dx, dy in directions:
                nx, ny = x + dx, y + dy

                if 0 <= nx < R and 0 <= ny < C and maze[nx][ny] != "+" and distance[nx][ny] == INF:
                    enqueue(nx, ny, d + 1)
                    if nx == 0 or nx == R - 1 or ny == 0 or ny == C - 1:
                        return d + 1

        return -1        

File: 1926-nearest-exit-from-entrance-in-maze/1926-nearest-exit-from-entrance-in-maze.cpp
#define inf 1000000000

class Solution {
public:
    
    void bfs_sssp(vector<vector<char>> &maze,int src_i,int src_j,vector<vector<int>> &dist){
        int i,j,m,n;
        m = maze.size();
        n = maze[0].size();
        
        queue<vector<int>> Q;
        
        Q.push({src_i,src_j});
        dist[src_i][src_j] = 0;
        while(Q.size() >= 1){
            i = (Q.front())[0];
            j = (Q.front())[1];
            Q.pop();
            
            // Up
            if( (i - 1) >= 0 && maze[i-1][j] == '.' && dist[i-1][j] == inf){
                Q.push({i-1,j});
                dist[i-1][j] = dist[i][j] + 1;
            }            
            // Down
            if((i+1) < m && maze[i+1][j] == '.' && dist[i+1][j] == inf){
                Q.push({i+1,j});
                dist[i+1][j] = dist[i][j] + 1;
            }            
            // Left
            if((j-1) >= 0 && maze[i][j-1] == '.' && dist[i][j-1] == inf){
                Q.push({i,j-1});
                dist[i][j-1] = dist[i][j] + 1;
            }
            // Right
            if((j+1) < n && maze[i][j+1] == '.' && dist[i][j+1] == inf){
                Q.push({i,j+1});
                dist[i][j+1] = (dist[i][j] + 1);
            }            
        }                
    }
    
    int nearestExit(vector<vector<char>>& maze, vector<int> &entrance){
        int m,n,i,j,src_i,src_j,ans;
        m = maze.size(), n = maze[0].size();
        
        src_i = entrance[0], src_j = entrance[1];
        
        // 1
        vector<vector<int>> dist(m,vector<int>(n,inf));
        bfs_sssp(maze,src_i,src_j,dist);
        
        // 2
        ans = inf;
        for(i=0 ; i<m ; i++){
            for(j=0 ; j<n ; j++){
                if(maze[i][j] == '.' && !(i==src_i && j == src_j) && (i == 0 || i == (m-1) || j == 0 || j == (n-1)) ){
                    ans = min(ans,dist[i][j]);
                }
            }
        }
                        
        return (ans == inf)? -1 : ans;            
    }
};